### PR TITLE
WP 331 `<Redirect>` component

### PR DESCRIPTION
**Pull request by mmcgahan**
This PR introduces a `<Redirect>` component that behaves similarly to [React Router's `<Redirect>`](https://reacttraining.com/react-router/web/api/Redirect) but includes support for redirecting to external URLs.

Docs have been added to `Routing.md` and the component itself, with full Flow type annotations. Unit tests and integration tests have been added.

### CODE REVIEW

**Comment by mattkime** [comment]
likely out of scope, but could something like this be its own package?

**Comment by mmcgahan** [comment]
@mattkime definitely - I think it will go nicely with the other Router-related tools into a proper `routing` module/package (mentioned in the [MWP modules issue](https://github.com/meetup/meetup-web-platform/issues/281)), and it would likely make sense to release it as a standalone repo/package at that point

File: .eslintrc.js [hunk 1]
```diff
@@ -34,15 +34,6 @@ const config = {
 		'no-useless-escape': 2,
 		'no-whitespace-before-property': 2,
 		'prefer-template': 2,
-		'require-jsdoc': [
```
Comment (mmcgahan) [inline .eslintrc.js]: At this point, we would prefer Flow over jsdoc for type annotations

File: package.json [hunk 2]
```diff
@@ -58,15 +58,6 @@
     "url-regex": "3.2.0",
     "uuid": "3.0.1"
   },
-  "peerDependencies": {
```
Comment (eilinora) [inline package.json]: are these supposed to all be removed?

Comment (mmcgahan) [inline package.json]: nope! had to remove them temporarily in order to run `flow-typed` but forgot to put them back, thanks!

File: flow-typed/platform.js [hunk 3]
```diff
@@ -71,3 +71,20 @@ declare module 'react-router-dom/matchPath' {
 	declare function matchPath(pathname: string, options: MatchPathOptions): null | Match;
 	declare module.exports: typeof matchPath;
 }
+
+declare type LocationShape = {
+  pathname?: string,
```
Comment (eilinora) [inline flow-typed/platform.js]: more a flow question what does the `?` mean?

Comment (mmcgahan) [inline flow-typed/platform.js]: optional prop. Any key that is not explicitly mentioned in the type definition will throw an error

File: src/components/redirect.test.jsx [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+	ReactDOMServer.renderToString(
+		<StaticRouter location={new URL('http://example.com')} context={context}>
+			<Redirect to={to} />
+		</StaticRouter>
+	);
+	return context;
+};
+
+describe('Server rendering', () => {
+	it('sets React Router context for "local" redirect, nothing from Redirect.rewind()', () => {
```
Comment (eilinora) [inline src/components/redirect.test.jsx]: is this because of withSideEffect?

File: src/components/redirect.test.jsx [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+		const internalTo = 'foo/bar';
+		const httpContext = renderToContext(internalTo);
+		expect(httpContext.url).toEqual(internalTo);
+		expect(Redirect.rewind()).toBeUndefined();
+	});
+	it('provides external redirect URL from Redirect.rewind(), nothing from React Router context', () => {
```
Comment (eilinora) [inline src/components/redirect.test.jsx]: nitpick alert! might be good to split this and the next `it` into separate tests

File: src/components/redirect.test.jsx (hunk 4 above)
Comment (mmcgahan) [inline src/components/redirect.test.jsx]: sort of. `Redirect.rewind()` is provided by `withSideEffect`, and should only return a value when the underlying React Router `<Redirect>` can't handle the URL - i.e. `rewind()` should only return a value when the Redirect is to an external URL. I'll add another inline comment [here](https://github.com/meetup/meetup-web-platform/pull/289/files/892b6732a97a369d4a6f005d8f7dd98f8af9b22e#diff-d88069df2a21944fc5327a27969009f7R65) to make it more explicit

File: src/components/redirect.test.jsx (hunk 5 above)
Comment (mmcgahan) [inline src/components/redirect.test.jsx]: You mean separate the https and http urls? I feel like that's not central to the behavior of the method, but I can split it out

File: src/components/Redirect.jsx [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+type RouterTo = string | LocationShape | URL;
+type RedirectProps = {
+	to: RouterTo,
+	push?: boolean,
+};
+
+/*
+ * A routing component that, when rendered, will redirect the application to
+ * another internal route _or_ external URL. On the server, the redirect will
+ * be handled with a 302 redirect HTTP response to the browser. On the client,
```
Comment (eilinora) [inline src/components/Redirect.jsx]: What if we need to do a 301? I'm soon going to be migrating a page that will live at 1 location to a new one, ideally i would use the Redirect to send people on their way but would be good to have it recorded as 301 so SEO picks that up

Comment (mmcgahan) [inline src/components/Redirect.jsx]: I started adding in that functionality this morning but it requires some non-trivial refactoring that I think would be easier to implement and review in a separate ticket - I'll do it today, but I'll make a new PR to keep this one simple and make it easier to see the code specifically relevant to 301/302 support

File: src/components/redirect.test.jsx (hunk 4 above)
Comment (mmcgahan) [inline src/components/redirect.test.jsx]: done

File: src/components/Redirect.jsx (hunk 6 above)
Comment (eilinora) [inline src/components/Redirect.jsx]: Sounds good! Thanks Mike!

**Pull request merged**